File: methods/zeros/b-newton/newton.c
#include <stdio.h>
#include <math.h>

void newton(double (*f)(double), double (*df)(double), double xo, int n)
{
    int iterations[] = {1, 3, 5};
    int k = 0;
    
    for (int i = 0; i < n; i++)
    {
        if (df(xo) == 0)
        {
            break;
        }

        double xi = xo - f(xo) / df(xo);

        if (i + 1 == iterations[k])
        {
            if(k)
            {
                printf(",\n");
            }

            printf("%.16f", xi);
            k++;
        }

        xo = xi;
    }
}

double f(double x)
{
    return sqrt(2 * 9.81 * x) * tanh(sqrt(2 * 9.81 * x) / (2 * 9.46) * 5.56) - 8.55;
}

double df(double x)
{
    return 2.88285 * pow(1/cosh(1.30168 * sqrt(x)),2) + (2.21472 * tanh(1.30168 * sqrt(x)))/sqrt(x);
}

int main()
{
    double xo = 0.16;
    int n = 5;

    newton(f, df, xo, n);

    return 0;
}

File: cunha/01-TRABALHO/newton.py
import math

def f(x):
    e = 2
    return x * (x - 1) * (x - 2) + 0.42 # function

def df(x):
    return 3*(x**2)-6*x+2 # derivada de f

def newton(x0,n):
    for i in range(0,n):
        x1 = x0-(f(x0)/df(x0))
        print(f'x_{i+1} = {x1:.7f}')
        x0=x1

def main():
    x0=2.7525393
    n=195
    newton(x0,n)

main()